app/src/ereferrals/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java

```diff
@@ -1,45 +1,169 @@
 package org.eyeseetea.malariacare.domain.usecase;
 
+import org.eyeseetea.malariacare.data.database.CredentialsLocalDataSource;
+import org.eyeseetea.malariacare.data.remote.OrgUnitDataSource;
 import org.eyeseetea.malariacare.domain.boundary.IAuthenticationManager;
+import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
+import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
+import org.eyeseetea.malariacare.domain.boundary.repositories.ICredentialsRepository;
+import org.eyeseetea.malariacare.domain.boundary.repositories.IOrgUnitRepository;
 import org.eyeseetea.malariacare.domain.entity.Credentials;
+import org.eyeseetea.malariacare.domain.entity.OrgUnit;
 import org.eyeseetea.malariacare.domain.entity.UserAccount;
 import org.eyeseetea.malariacare.domain.exception.InvalidCredentialsException;
 import org.eyeseetea.malariacare.domain.exception.NetworkException;
+import org.eyeseetea.malariacare.domain.exception.PullConversionException;
+import org.eyeseetea.malariacare.network.ServerAPIController;
+import org.eyeseetea.malariacare.presentation.executors.AsyncExecutor;
+import org.eyeseetea.malariacare.presentation.executors.UIThreadExecutor;
 
 import java.net.MalformedURLException;
 import java.net.UnknownHostException;
 
 public class LoginUseCase extends ALoginUseCase {
     private IAuthenticationManager mAuthenticationManager;
+    private Credentials insertedCredentials;
+    private IMainExecutor mMainExecutor = new UIThreadExecutor();
+    private IAsyncExecutor mAsyncExecutor = new AsyncExecutor();
+    private Callback mCallback;
 
-    public LoginUseCase(IAuthenticationManager authenticationManager) {
+    public LoginUseCase(IAuthenticationManager authenticationManager, IMainExecutor mainExecutor,
+            IAsyncExecutor asyncExecutor) {
         mAuthenticationManager = authenticationManager;
+        mMainExecutor = mainExecutor;
+        mAsyncExecutor = asyncExecutor;
     }
 
     @Override
-    public void execute(Credentials credentials, final Callback callback) {
-        mAuthenticationManager.login(credentials,
+    public void execute(final Credentials credentials, final Callback callback) {
+        mCallback = callback;
```

Comment:
@manuelplazaspalacio All steps of login process should be executed with AsyncExecutor first step included. Review PushUseCase for view a example of when execute in async mode